ontology/obo/tree.go

```diff
@@ -0,0 +1,160 @@
+package obo
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"log"
+)
+
+// BuildTree takes a slice of Obo structs and builds a tree, setting
+// the Parents and Children fields to point to related nodes.
+func BuildTree(terms []Obo) map[string]*Obo {
+	termMap := makeTermMap(terms)
+
+	var parentId string
+	var foundInMap bool
+	var parentTerm *Obo
+	var isAItem IsADescription
+
+	// Now we iterate through terms and set up the tree
+	for i := range terms {
+		term := &terms[i] //get a memory copy
+		for _, isAItem = range term.IsA {
+			parentId = isAItem.ParentId
+			if parentTerm, foundInMap = termMap[parentId]; foundInMap {
+				term.Parents = append(term.Parents, parentTerm)
+				parentTerm.Children = append(parentTerm.Children, term)
+			} else {
+				//log.Fatalf("Error: The term with ID \"%s\" has a parent with ID \"%s\" that "+
+				//	"is not found in the Obo file.\n", term.Id, parentId)
+			}
+		}
+	}
+	return termMap
+}
+
+// termToDot is a helper function of ToDot that converts and individual Obo struct to a
+// DOT format digraph descriptor, and recursively does so for all children of the input Obo struct.
+func termToDot(term Obo, out *fileio.EasyWriter, visitedNode map[string]bool) {
+	var err error
+	if visitedNode[term.Id] {
+		return
+	}
+	visitedNode[term.Id] = true
+	_, err = fmt.Fprintf(out, "\"%s\" [label = \"%s\"];\n", term.Id, term.Name)
+	exception.PanicOnErr(err)
+
+	// now we recursively visit children
+	for _, child := range term.Children {
+		_, err = fmt.Fprintf(out, "\"%s\" -> \"%s\";\n", term.Id, child.Id)
+		termToDot(*child, out, visitedNode)
+	}
+}
+
+// ToDot formats an input Obo slice as a tree and writes the resulting tree to a file in DOT format.
+func ToDot(outFile string, terms []Obo) {
+	var err error
+	out := fileio.EasyCreate(outFile)
+	BuildTree(terms)
+
+	//write header line for Dot
+	_, err = fmt.Fprintf(out, "digraph G{\n")
+	exception.PanicOnErr(err)
+
+	var visitedNode = make(map[string]bool)
+
+	for _, term := range terms {
+		termToDot(term, out, visitedNode)
+	}
+
+	//write tail line for Dot and close file
+	_, err = fmt.Fprintf(out, "}\n")
+	err = out.Close()
+	exception.PanicOnErr(err)
+}
+
+// makeTermMap creates a map to find nodes based on their ID and AltIDs
+func makeTermMap(terms []Obo) map[string]*Obo {
+	var altId string
+	termMap := make(map[string]*Obo)
+	for i := range terms {
+		term := &terms[i] // get a memory copy
```

Comment:
I think this function should take a pointer to the root and then use recursion to make a map of pointers to the nodes.  I don't think we want/need the copying